tests/integration/utils/flow_tools.py

```diff
@@ -54,13 +54,24 @@ def _make_flow(run_dir, test_dir, conf, name=None):
     return reg
 
 
-def _make_scheduler(reg, **opts):
+@contextmanager
+def _make_scheduler():
     """Return a scheduler object for a flow registration."""
-    # This allows paused_start to be overriden:
-    opts = {'paused_start': True, **opts}
-    options = RunOptions(**opts)
-    # create workflow
-    return Scheduler(reg, options)
+    schd: Scheduler = None  # type: ignore
+
+    def __make_scheduler(reg: str, **opts: Any) -> Scheduler:
+        # This allows paused_start to be overriden:
+        opts = {'paused_start': True, **opts}
+        options = RunOptions(**opts)
+        # create workflow
+        nonlocal schd
+        schd = Scheduler(reg, options)
+        return schd
+
+    yield __make_scheduler
+    # Teardown
+    if hasattr(schd, 'workflow_db_mgr'):
+        schd.workflow_db_mgr.on_workflow_shutdown()
```

Comment:
Does this mean that `on_workflow_shutdown` isn't being called?

We shouldn't need any teardown logic here as the `Scheduler` is supposed to handle all of this itself (if it isn't that's a bug).

Reply:
Integration tests like this don't need it:
https://github.com/cylc/cylc-flow/blob/0230c9da0690ead76e9747a8f2f16bf175f3d51b/tests/integration/test_examples.py#L59-L60

It's tests like this that need the teardown:
https://github.com/cylc/cylc-flow/blob/0230c9da0690ead76e9747a8f2f16bf175f3d51b/tests/integration/test_examples.py#L137-L140

because they don't start the scheduler, and so don't shut it down either

Reply:
Ah, dammit, we are creating the DB in order to write the workflow parameters to it at configure time so the `on_workflow_start` method is really `on_workflow_configure`. My bad, I made a rough cut at pulling apart the different stages of workflow startup with the original `async` work but they are still quite tangled and need a bit of TLC.

I've had a quick try at delaying the DB creation until workflow start, this diff passes `tests/i`:

```diff
diff --git a/cylc/flow/scheduler.py b/cylc/flow/scheduler.py
index 06b9de96c..46d662ea5 100644
--- a/cylc/flow/scheduler.py
+++ b/cylc/flow/scheduler.py
@@ -430,8 +430,6 @@ class Scheduler:
         self.load_flow_file()
         self.profiler.log_memory("scheduler.py: after load_flow_file")
 
-        self.workflow_db_mgr.on_workflow_start(self.is_restart)
-
         if not self.is_restart:
             # Set workflow params that would otherwise be loaded from database:
             self.options.utc_mode = get_utc_mode()
@@ -463,6 +461,7 @@ class Scheduler:
 
         self.profiler.log_memory("scheduler.py: before load_tasks")
         if self.is_restart:
+            self.workflow_db_mgr.on_workflow_restart_configure()
             self.load_tasks_for_restart()
             if self.restored_stop_task_id is not None:
                 self.pool.set_stop_task(self.restored_stop_task_id)
@@ -471,10 +470,6 @@ class Scheduler:
         self.process_cylc_stop_point()
         self.profiler.log_memory("scheduler.py: after load_tasks")
 
-        self.workflow_db_mgr.put_workflow_params(self)
-        self.workflow_db_mgr.put_workflow_template_vars(self.template_vars)
-        self.workflow_db_mgr.put_runtime_inheritance(self.config)
-
         self.already_timed_out = False
         self.set_workflow_timer()
 
@@ -567,6 +562,10 @@ class Scheduler:
 
     async def start_scheduler(self):
         """Start the scheduler main loop."""
+        self.workflow_db_mgr.on_workflow_start(self.is_restart)
+        self.workflow_db_mgr.put_workflow_params(self)
+        self.workflow_db_mgr.put_workflow_template_vars(self.template_vars)
+        self.workflow_db_mgr.put_runtime_inheritance(self.config)
         try:
             self._configure_contact()
             if self.is_restart:
diff --git a/cylc/flow/workflow_db_mgr.py b/cylc/flow/workflow_db_mgr.py
index f471d0c72..e04f7038a 100644
--- a/cylc/flow/workflow_db_mgr.py
+++ b/cylc/flow/workflow_db_mgr.py
@@ -181,6 +181,10 @@ class WorkflowDatabaseManager:
         else:
             return json.dumps([type(obj).__name__, obj.__getnewargs__()])
 
+    def on_workflow_restart_configure(self):
+        self.pri_dao = self.get_pri_dao()
+        os.chmod(self.pri_path, 0o600)
+
     def on_workflow_start(self, is_restart):
         """Initialise data access objects.
 
@@ -194,7 +198,8 @@ class WorkflowDatabaseManager:
             except OSError:
                 # Just in case the path is a directory!
                 rmtree(self.pri_path, ignore_errors=True)
-        self.pri_dao = self.get_pri_dao()
+        if not self.pri_dao:
+            self.pri_dao = self.get_pri_dao()
         os.chmod(self.pri_path, 0o600)
         self.pub_dao = CylcWorkflowDAO(self.pub_path, is_public=True)
         self.copy_pri_to_pub()
```

Reply:
Ok yeah, I've started along this path based on that diff but it's starting to balloon, so I'll put the rest in a later PR